File: src/jackett.ts
import { XMLParser } from "fast-xml-parser";
import { getReqEnvVariable } from "./loadenv.ts";

const JACKETT_URL = getReqEnvVariable("JACKETT_URL");
const JACKETT_API_KEY = getReqEnvVariable("JACKETT_API_KEY");

if (!JACKETT_URL) {
  throw new Error("JACKETT_URL environment variable is required.");
}
if (!JACKETT_API_KEY) {
  throw new Error("JACKETT_API_KEY environment variable is required.");
}

export interface JackettResult {
  Title: string;
  Tracker: string;
  TrackerId: string;
  Link?: string;
  MagnetUri?: string;
  InfoHash?: string;
  Seeders: number;
  Peers: number;
  Size: number;
}

export type TorznabParams =
  | { type: "movie_imdb"; imdbId: string }
  | { type: "movie_text"; query: string }
  | { type: "series_imdb"; imdbId: string; season: number; episode: number }
  | { type: "series_text"; query: string; season: number; episode: number }
  | { type: "series_season_imdb"; imdbId: string; season: number }
  | { type: "series_season_text"; query: string; season: number };

function buildTorznabQueryString(params: TorznabParams): string {
  const queryParams = new URLSearchParams();
  queryParams.append("apikey", JACKETT_API_KEY as string);

  switch (params.type) {
    case "movie_imdb":
      queryParams.append("t", "movie");
      queryParams.append("imdbid", params.imdbId);
      break;
    case "movie_text":
      queryParams.append("t", "movie");
      queryParams.append("q", params.query);
      break;
    case "series_imdb":
      queryParams.append("t", "tvsearch");
      queryParams.append("imdbid", params.imdbId);
      queryParams.append("season", params.season.toString());
      queryParams.append("ep", params.episode.toString());
      break;
    case "series_text":
      queryParams.append("t", "search");
      queryParams.append(
        "q",
        `${params.query} S${params.season.toString().padStart(2, "0")}E${params.episode.toString().padStart(2, "0")}`,
      );
      break;
    case "series_season_imdb":
      queryParams.append("t", "tvsearch");
      queryParams.append("imdbid", params.imdbId);
      queryParams.append("season", params.season.toString());
      break;
    case "series_season_text":
      queryParams.append("t", "search");
      queryParams.append(
        "q",
        `${params.query} S${params.season.toString().padStart(2, "0")}`,
      );
      break;
  }

  return queryParams.toString();
}

export async function fetchJackettResults(
  torznabParams: TorznabParams,
): Promise<JackettResult[]> {
  const qs = buildTorznabQueryString(torznabParams);
  const url = `${JACKETT_URL}/api/v2.0/indexers/all/results/torznab/api?${qs}`;

  try {
    const response = await fetch(url);
    if (!response.ok) {
      console.error(`Jackett API responded with status: ${response.status}`);
      return [];
    }

    const xml = await response.text();
    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: "@_",
      isArray: (name) => name === "item" || name === "torznab:attr",
    });
    const jsonObj = parser.parse(xml);
    const items = jsonObj.rss?.channel?.item || [];
    const results: JackettResult[] = [];

    for (const item of items) {
      const title = item.title ? String(item.title) : "Unknown";
      const size = parseInt(item.size, 10) || 0;
      const link = item.link;

      const enclosureUrl = item.enclosure && item.enclosure["@_url"];
      const magnetUri =
        enclosureUrl && enclosureUrl.startsWith("magnet:")
          ? enclosureUrl
          : link;

      let seeders = 0;
      let infoHash = undefined;

      const attrs = item["torznab:attr"] || [];
      for (const attr of attrs) {
        if (attr["@_name"] === "seeders") {
          seeders = parseInt(attr["@_value"], 10) || 0;
        }
        if (attr["@_name"] === "infohash") {
          infoHash = attr["@_value"];
        }
      }

      let trackerId = "";
      let trackerName = "Unknown";

      if (item.jackettindexer) {
        if (typeof item.jackettindexer === "object") {
          trackerId = item.jackettindexer["@_id"] || "";
          trackerName = item.jackettindexer["#text"] || "Unknown";
        } else {
          trackerName = String(item.jackettindexer);
        }
      }

      if (seeders > 0) {
        results.push({
          Title: title,
          Tracker: trackerName,
          TrackerId: trackerId,
          Link: magnetUri,
          MagnetUri: magnetUri,
          InfoHash: infoHash,
          Seeders: seeders,
          Peers: 0,
          Size: size,
        });
      }
    }

    return results.sort((a, b) => (b.Seeders || 0) - (a.Seeders || 0));
  } catch (error) {
    console.error("Failed to fetch from Jackett API:", error);
    return [];
  }
}

File: src/jackett.rs
use crate::config::Config;
use log::{debug, error};
use reqwest::Client;
use serde::Deserialize;

#[derive(Debug, Clone)]
pub struct JackettResult {
    pub title: String,
    pub tracker: String,
    // pub tracker_id: String,
    pub link: Option<String>,
    pub magnet_uri: Option<String>,
    pub info_hash: Option<String>,
    pub seeders: u32,
    pub size: u64,
}

pub enum TorznabParams {
    MovieImdb { imdb_id: String },
    MovieText { query: String },
    SeriesImdb { imdb_id: String, season: u32, episode: u32 },
    SeriesText { query: String, season: u32, episode: u32 },
    SeriesSeasonImdb { imdb_id: String, season: u32 },
    SeriesSeasonText { query: String, season: u32 },
}

fn build_query_string(config: &Config, params: &TorznabParams) -> String {
    let mut qs = format!("apikey={}", config.jackett_api_key);
    match params {
        TorznabParams::MovieImdb { imdb_id } => {
            qs.push_str(&format!("&t=movie&imdbid={}", imdb_id));
        }
        TorznabParams::MovieText { query } => {
            qs.push_str(&format!("&t=movie&q={}", urlencoding::encode(query)));
        }
        TorznabParams::SeriesImdb { imdb_id, season, episode } => {
            qs.push_str(&format!("&t=tvsearch&imdbid={}&season={}&ep={}", imdb_id, season, episode));
        }
        TorznabParams::SeriesText { query, season, episode } => {
            let q = format!("{} S{:02}E{:02}", query, season, episode);
            qs.push_str(&format!("&t=search&q={}", urlencoding::encode(&q)));
        }
        TorznabParams::SeriesSeasonImdb { imdb_id, season } => {
            qs.push_str(&format!("&t=tvsearch&imdbid={}&season={}", imdb_id, season));
        }
        TorznabParams::SeriesSeasonText { query, season } => {
            let q = format!("{} S{:02}", query, season);
            qs.push_str(&format!("&t=search&q={}", urlencoding::encode(&q)));
        }
    }
    qs
}

// Quick-XML serde structs
#[derive(Debug, Deserialize)]
struct Rss {
    channel: Option<Channel>,
}

#[derive(Debug, Deserialize)]
struct Channel {
    #[serde(rename = "item", default)]
    items: Vec<Item>,
}

#[derive(Debug, Deserialize)]
struct Item {
    title: Option<String>,
    size: Option<u64>,
    link: Option<String>,
    enclosure: Option<Enclosure>,
    #[serde(rename = "attr", default)]
    attrs: Vec<Attr>,
    jackettindexer: Option<JackettIndexer>,
}

#[derive(Debug, Deserialize)]
struct Enclosure {
    #[serde(rename = "@url")]
    url: Option<String>,
}

#[derive(Debug, Deserialize)]
struct Attr {
    #[serde(rename = "@name")]
    name: Option<String>,
    #[serde(rename = "@value")]
    value: Option<String>,
}

#[derive(Debug, Deserialize)]
struct JackettIndexer {
    // #[serde(rename = "@id")]
    // id: Option<String>,
    #[serde(rename = "$value")]
    name: Option<String>,
}

pub async fn fetch_jackett_results(
    client: &Client,
    config: &Config,
    params: &TorznabParams,
) -> Vec<JackettResult> {
    let qs = build_query_string(config, params);
    let url = format!("{}/api/v2.0/indexers/all/results/torznab/api?{}", config.jackett_url, qs);

    debug!("Fetching from Jackett: {}", url);
    let response = match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r,
        Ok(r) => {
            error!("Jackett API responded with status: {}", r.status());
            return vec![];
        }
        Err(e) => {
            error!("Failed to fetch from Jackett API: {}", e);
            return vec![];
        }
    };

    let xml = match response.text().await {
        Ok(t) => t,
        Err(_) => return vec![],
    };

    let rss: Rss = match quick_xml::de::from_str(&xml) {
        Ok(r) => r,
        Err(e) => {
            error!("Failed to parse Jackett XML: {}", e);
            error!("Raw XML was: {}", xml);
            return vec![];
        }
    };

    let mut results = Vec::new();

    if let Some(channel) = rss.channel {
        if channel.items.is_empty() {
            debug!("Jackett returned a channel, but no items. Raw XML: {}", xml);
        }
        for item in channel.items {
            let title = item.title.unwrap_or_else(|| "Unknown".to_string());
            let size = item.size.unwrap_or(0);
            let link = item.link.clone();

            let enclosure_url = item.enclosure.and_then(|e| e.url);
            let magnet_uri = if let Some(url) = enclosure_url.clone() {
                if url.starts_with("magnet:") {
                    Some(url)
                } else {
                    link.clone().or(Some(url))
                }
            } else {
                link.clone()
            };

            let mut seeders = 0;
            let mut info_hash = None;

            for attr in item.attrs {
                if let (Some(name), Some(value)) = (attr.name, attr.value) {
                    if name == "seeders" {
                        seeders = value.parse().unwrap_or(0);
                    } else if name == "infohash" {
                        info_hash = Some(value);
                    }
                }
            }

            // Fallback: Try to extract info_hash from magnet_uri if missing
            if info_hash.is_none() {
                if let Some(ref magnet) = magnet_uri {
                    if magnet.starts_with("magnet:") {
                        if let Some(start) = magnet.find("urn:btih:") {
                            let hash_part = &magnet[start + 9..];
                            let end = hash_part.find('&').unwrap_or(hash_part.len());
                            info_hash = Some(hash_part[..end].to_string());
                        }
                    }
                }
            }

            // let mut tracker_id = String::new();
            let mut tracker_name = "Unknown".to_string();

            if let Some(indexer) = item.jackettindexer {
                // tracker_id = indexer.id.unwrap_or_default();
                tracker_name = indexer.name.unwrap_or_else(|| "Unknown".to_string());
            }

            if seeders > 0 {
                results.push(JackettResult {
                    title,
                    tracker: tracker_name,
                    // tracker_id,
                    link,
                    magnet_uri,
                    info_hash,
                    seeders,
                    size,
                });
            }
        }
        debug!("Jackett XML parsed successfully. Found {} items.", results.len());
    } else {
        debug!("Jackett XML did not contain a <channel>. Raw XML: {}", xml);
    }

    results.sort_by(|a, b| b.seeders.cmp(&a.seeders));
    results
}
